lib/services/local/saml_idp_service_provider.go

```diff
@@ -295,3 +307,99 @@ func (s *SAMLIdPServiceProviderService) generateAndSetEntityDescriptor(sp types.
+	if reflect.DeepEqual(sp.GetAttributeMapping(), embeddedAttributes) {
+		return nil
+	}
+
+	newEmbeddedSPSSODescriptor := spSSODescriptorWithRequestedAttributes(sp.GetAttributeMapping())
+	// If there is an existing embedded SPSSODescriptor, we'll replace it with newEmbeddedSPSSODescriptor
+	// to avoid duplication or possible fragmented SPSSODescriptor. This is checked with embeddedSPSSODescriptorIndex
+	// value greater than 0. Otherwise, simply append to ed.SPSSODescriptors.
+	if embeddedSPSSODescriptorIndex == 0 {
+		ed.SPSSODescriptors = append(ed.SPSSODescriptors, newEmbeddedSPSSODescriptor)
+	} else {
+		ed.SPSSODescriptors[embeddedSPSSODescriptorIndex] = newEmbeddedSPSSODescriptor
+	}
+
+	edWithAttributes, err := xml.MarshalIndent(ed, " ", "    ")
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	s.log.Infof("attribute mapping embedded in entity descriptor")
+	sp.SetEntityDescriptor(string(edWithAttributes))
+	return nil
+}
+
+// spSSODescriptorWithRequestedAttributes builds new saml.SPSSODescriptor populated with
+// types.SAMLAttributeMapping (attributeMapping input) converted to saml.RequestedAttributes format.
+func spSSODescriptorWithRequestedAttributes(attributeMapping []*types.SAMLAttributeMapping) saml.SPSSODescriptor {
+	var reqs []saml.RequestedAttribute
+	for _, v := range attributeMapping {
+		var reqAttrs saml.RequestedAttribute
+		reqAttrs.FriendlyName = v.Name
+		reqAttrs.Name = v.Name
+		reqAttrs.NameFormat = v.NameFormat
+		reqAttrs.Values = append(reqAttrs.Values, saml.AttributeValue{
+			Value: v.Value,
+		})
+		reqs = append(reqs, reqAttrs)
+	}
+
+	return saml.SPSSODescriptor{
+		AttributeConsumingServices: []saml.AttributeConsumingService{
+			{
+				// ServiceNames is hardcoded with value TELEPORT_SAML_IDP to make the descriptor
+				// recognizable throughout SAML SSO flow. Attribute mapping should only ever
+				// edit SPSSODescriptor containing TELEPORT_SAML_IDP element. Otherwise, we risk
+				// overriding SP managed SPSSODescriptor!
+				ServiceNames:        []saml.LocalizedName{{Value: teleportSAMLIdP}},
+				RequestedAttributes: reqs,
+			},
+		},
+	}
+}
+
+// samlSPSSODescriptorToProtoSAMLAttributeMapping converts []saml.SPSSODescriptor to []*types.SAMLAttributeMapping
+// and returns SPSSODescriptor element index and converted attribute mapping values.
+// The correct SPSSODescriptor is determined by searching for AttributeConsumingService element with ServiceNames named
+// TELEPORT_SAML_IDP.
+func samlSPSSODescriptorToProtoSAMLAttributeMapping(spSSODescriptors []saml.SPSSODescriptor) (int, []*types.SAMLAttributeMapping) {
+	embeddedSPSSODescriptorIndex := 0
+	attrs := make([]*types.SAMLAttributeMapping, 0)
+	for descriptorIndex, descriptor := range spSSODescriptors {
+		for _, acs := range descriptor.AttributeConsumingServices {
+			for _, serviceName := range acs.ServiceNames {
+				if serviceName.Value == teleportSAMLIdP {
```

Comment:
Instead of building the staircase of loops in this function, let's make a helper method that returns Teleport's descriptor and index given a list of descriptors.

Reply:
While I don't like it myself, I am not sure if we could get away with that nested loop entirely (at least from what I can think of my head🤔) cause the type `[]saml.SPSSODescriptor` comes from external package and all of it is nested slices